probe-rs/src/probe/wlink/mod.rs

```diff
@@ -0,0 +1,606 @@
+        match self {
+            WchLinkVariant::Ch549 => write!(f, "WCH-Link-CH549"),
+            WchLinkVariant::ECh32v307 => write!(f, "WCH-LinkE-CH32V307"),
+            WchLinkVariant::SCh32v203 => write!(f, "WCH-LinkS-CH32V203"),
+            WchLinkVariant::B => write!(f, "WCH-LinkB"),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+pub enum RiscvChip {
+    CH32V103 = 0x01,
+    CH573 = 0x02,
+    CH569 = 0x03,
+    CH32V20x = 0x05,
+    CH32V30x = 0x06,
+    CH32V003 = 0x09,
+}
+
+impl RiscvChip {
+    fn from_u8(value: u8) -> Option<Self> {
+        match value {
+            0x01 => Some(RiscvChip::CH32V103),
+            0x02 => Some(RiscvChip::CH573),
+            0x03 => Some(RiscvChip::CH569),
+            0x05 => Some(RiscvChip::CH32V20x),
+            0x06 => Some(RiscvChip::CH32V30x),
+            0x09 => Some(RiscvChip::CH32V003),
+            _ => None,
+        }
+    }
+}
+
+/// WCH-Link device (mod:RV)
+#[derive(Debug)]
+pub(crate) struct WchLink {
+    device: WchLinkUsbDevice,
+    name: String,
+    variant: WchLinkVariant,
+    v_major: u8,
+    v_minor: u8,
+    riscvchip: u8,
+    chip_type: u32,
+    // Hack to support NOP after READ
+    last_dmi_read: Option<(u8, u32, u8)>,
+}
+
+impl WchLink {
+    fn get_version(&mut self) -> Result<(), DebugProbeError> {
+        tracing::debug!("Getting version of WCH-Link...");
+
+        let rxbuf = self
+            .device
+            .write_command(&[0x81, 0x0d, 0x01, 0x01], Duration::from_millis(300))?;
+
+        if rxbuf[0] != 0x82 {
+            return Err(WchLinkError::CommandFailed(rxbuf[1]).into());
+        }
+
+        if rxbuf.len() == 5 {
+            self.variant = WchLinkVariant::Ch549;
+            self.v_major = rxbuf[3];
+            self.v_minor = rxbuf[4];
+            tracing::warn!(
+                "WCH-Link {} {}.{} is outdated, please update the firmware.",
+                self.variant,
+                self.v_major,
+                self.v_minor
+            );
+        } else if rxbuf.len() == 6 {
+            self.variant = WchLinkVariant::Ch549;
+            self.v_major = rxbuf[3];
+            self.v_minor = rxbuf[4];
+            tracing::warn!(
+                "WCH-Link {} {}.{} is outdated, please update the firmware.",
+                self.variant,
+                self.v_major,
+                self.v_minor
+            );
+        } else if rxbuf.len() == 7 {
+            self.v_major = rxbuf[3];
+            self.v_minor = rxbuf[4];
+
+            match rxbuf[5] {
+                1 => self.variant = WchLinkVariant::Ch549,
+                2 => self.variant = WchLinkVariant::ECh32v307,
+                3 => self.variant = WchLinkVariant::SCh32v203,
+                4 => self.variant = WchLinkVariant::B,
+                _ => return Err(WchLinkError::UnknownDevice.into()),
+            }
+        } else {
+            unimplemented!("unknown rx packet length {}", rxbuf.len());
```

Comment:
Please return an error here, `unimplemented!()` should not be used for this case.